ros2_generate_interface_docs/ros2_generate_interface_docs/utils.py

```diff
@@ -0,0 +1,153 @@
+
+from io import StringIO
+import os
+import sys
+
+import em
+
+from rosidl_parser.definition import AbstractNestedType, BASIC_TYPES, NamespacedType
+
+_TEMPLATES_DIR = 'templates'
+IGNORED_KEYS = ['__slots__', '__doc__', '_fields_and_field_types', 'SLOT_TYPES', '__init__',
+                '__repr__', '__eq__', '__hash__', 'get_fields_and_field_types', '__module__']
+
+
+def generate_raw_text(raw_text):
+    raw_test_str = ''
+    for line in raw_text.split(os.linesep):
+        line = line.replace(' ', '&nbsp;')
+        parts = line.split('#')
+        if len(parts) > 1:
+            raw_test_str = raw_test_str + parts[0] \
+                + '<font color="blue">#%s</font><br/>' % ('#'.join(parts[1:]))
+        else:
+            raw_test_str = raw_test_str + '%s<br/>' % parts[0]
+    return raw_test_str
+
+
+def resource_name(resource):
+    """Return the resource name."""
+    if '/' not in resource:
+        return '', '', resource
+    values = resource.split('/')
+    if len(values) != 3:
+        raise ValueError(...)
+    return tuple(values)
+
+
+def get_templates_dir():
+    """Return template directory."""
+    return os.path.join(os.path.dirname(__file__), _TEMPLATES_DIR)
+
+
+def copy_css_style(folder_name):
+    """Copy the css style file in the folder_name."""
+    style_css, _ = load_template('styles.css')
+    with open(os.path.join(folder_name, 'styles.css'), 'w') as f:
+        f.write(style_css)
+
+    msg_style_css, _ = load_template('msg-styles.css')
+    with open(os.path.join(folder_name, 'msg-styles.css'), 'w') as f:
+        f.write(msg_style_css)
+
+
+def load_template(filename):
+    """
+    Look up file within rosdoc ROS package.
+
+    return its content, may sys.exit on error.
+    Contents are cached with filename as key.
+
+    :returns: cached file contents
+    """
+    filename = os.path.join(get_templates_dir(), filename)
+    if not os.path.isfile(filename):
+        sys.stderr.write("Cannot locate template file '%s'\n" % (filename))
+        sys.exit(1)
+    with open(filename, 'r') as f:
+        content = f.read()
+        if not content:
+            raise Exception("Template file '%s' is empty\n" % (filename))
+        return content, filename
+
+
+def write_template(template_name, data, output_file):
+    msg_index_template, template_path = load_template(template_name)
+    output = StringIO()
+    interpreter = em.Interpreter(
+        output=output,
+        options={
+            em.BUFFERED_OPT: True,
+            em.RAW_OPT: True,
+        },
+    )
+    try:
+        with open(template_path, 'r') as h:
+            template_content = h.read()
+            interpreter.invoke(
+                'beforeFile', name=template_name, file=h, locals=data)
+        interpreter.string(template_content, template_path, locals=data)
+        interpreter.invoke('afterFile')
+    except Exception as e:  # noqa: F841
+        if os.path.exists(output_file):
+            os.remove(output_file)
```

Comment:
What do you think about leaving this file in place in case the user wants to debug the issue that causes this exception to be thrown?

Reply:
6ced44d4b6ea365a6c4128027ac0173f10b6ec12